clippy_lints/src/use_last.rs

```diff
@@ -0,0 +1,123 @@
+/// **Why is this bad?** Using `x.last()` is easier to read and has the same result.
+///
+/// Note that using `x[x.len() - 1]` is semantically different from `x.last()`.
+/// Indexing into the array will panic on out-of-bounds accesses, while
+/// `x.get()` and `x.last()` will return `None`.
+///
+/// There is another lint (get_unwrap) that covers the case of using
+/// `x.get(index).unwrap()` instead of `x[index]`.
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+///
+/// ```rust
+/// // Bad
+/// let x = vec![2, 3, 5];
+/// let last_element = x.get(x.len() - 1);
+///
+/// // Good
+/// let x = vec![2, 3, 5];
+/// let last_element = x.last();
+/// ```
+
+declare_clippy_lint! {
+    pub USE_LAST,
+    complexity,
+    "using `x.get(x.len() - 1)` instead of `x.last()`"
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct UseLast;
+
+impl LintPass for UseLast {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(USE_LAST)
+    }
+
+    fn name(&self) -> &'static str {
+        "UseLast"
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UseLast {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        if_chain! {
+            // let _ = println!("Starting UseLast");
+            // Is a method call
+            if let ExprKind::MethodCall(ref path, _, ref args) = expr.node;
+            // let _ = println!("It is a MethodCall");
+
+            // Method name is "get"
+            if path.ident.name == "get";
+            // let _ = println!("The name is get");
+
+            // Argument 0 (the struct we're calling the method on) is a vector
+            if let Some(struct_calling_on) = args.get(0);
+            // let _ = println!("It had an argument");
+            let struct_ty = cx.tables.expr_ty(struct_calling_on);
+            if match_type(cx, struct_ty, &paths::VEC);
+            // let _ = println!("It was a vector");
+
+            // Argument to "get" is a binary operation
+            if let Some(get_index_arg) = args.get(1);
+            // let _ = println!("It had an argument");
+            if let rustc::hir::ExprKind::Binary(ref op, ref lhs, ref rhs) = get_index_arg.node;
+            // let _ = println!("It was a vector");
+
+            // Binary operation is a subtraction
+            if op.node == rustc::hir::BinOpKind::Sub;
+            // let _ = println!("It was a subtraction");
+
+            // LHS of subtraction is "x.len()"
+            if let ExprKind::MethodCall(ref arg_lhs_path, _, ref lhs_args) = lhs.node;
+             // let _ = println!("LHS of sub is a method call");
+            if arg_lhs_path.ident.name == "len";
+            // let _ = println!("LHS of sub was method named len");
+            if let Some(_arg_lhs_struct) = lhs_args.get(0);
+            // let _ = println!("LHS of sub method has an arg");
+
+            // TODO: Is this a valid way to check if they reference the same vector?
+            // if arg_lhs_struct.hir_id == struct_calling_on.hir_id;
```

Comment:
you need to check whether the two expressions are equal by structure. I think we have an `expr_eq` function or similar code somewhere in clippy.